### ISSUE TITLE
`makedns`/`makehost` to support same hostname in different domains

### ISSUE THREAD

**Issue by kcgthb** [xcat2/xcat-core#2142]
To followup on #1086, we'd like to have a way to generate `/etc/hosts` and DNS records with the same hostname in different domains.

The requirement would be to have:

192.168.0.1 hosta.private.zone hosta
172.16.0.1   hosta.public.com
in /etc/hosts, ie. the same hostname (hosta) without any suffix in 2 different domains (.private.zone and .public.com).

As of xCAT 2.12.2, it doesn't seem possible to generate identical hostnames in different DNS zones. Would it be possible to implement such a feature?

**Comment by bybai** [comment]
Hi @kcgthb ,
In my pull request  #1902, it is a litte different with you request here.
I added suffix and the same hostname in different domains.
Could you tell me why you do not add any suffix ?

**Comment by kcgthb** [comment]
Hi @bybai 

Sure! Our requirement is to have the same short hostname (the node name) resolve in multiple networks. That's because some servers in our environment only have access to specific networks, but we need to be able to reference multi-homes nodes with the same unique name, wherever we're connecting from.

For instance, let's say we have 2 networks, `net1` and `net2`, and 3 hosts:
* `host1` has a single interface in `net1`
* `host2` has a single interface in `net2`
* `host3` has an interface in both `net1` and `net2`.

We want to be able to reference `host3` as `host3.net1` from `host1` and as `host3.net2` from `host2`. Actually, since `net1` will be the default domain in `host1`'s `resolv.conf` (and `net2` in `host2`'s), we'd be able to just use `host3` from both `host1` and `host2`.

Having a suffix, even if it's the same on both nets, is not convenient and seems pretty artificial. We need to be able to refer to each host using their hostnames, not an alias.

**Comment by bybai** [comment]
Hi @kcgthb ,
This requirement is added in my list, Have discussed a better user interface for both #2142 and #1086.

**Comment by kcgthb** [comment]
Hi @bybai 

I was wondering if you had any news about this? I would really be able to have the ability to define nodes with the same short hostname on different interfaces/networks.

Thanks!

**Comment by bybai** [comment]
Hi @kcgthb ,
I start to work on this now, thanks.

**Comment by kcgthb** [comment]
Hi @bybai 

I think I actually found a workaround for this. Defining an extra nic alias for the interface in the `nics` table that is the short hostname will provide the expected result:

For instance, given the following objects (note how `nicaliases.bmc` is set to the node name)
```
# lsdef sh-sl01 --nics
Object name: sh-sl01
    nicaliases.bmc=sh-sl01
    nicips.bmc=10.11.0.11
    nicnetworks.bmc=net_ipmi
    nictypes.bmc=bmc
# lsdef -t network net_ipmi
Object name: net_ipmi
    domain=ipmi
    gateway=10.11.0.99
    mask=255.255.0.0
    mgtifname=em1:0
    net=10.11.0.0
```
then the generated `/etc/hosts` contains an alias for the BMC interface which is set to the short hostname:
```
# makehost sh-sl01
# grep sh-sl01-bmc /etc/hosts
10.11.0.11 sh-sl01-bmc.ipmi sh-sl01-bmc sh-sl01
```
and `makedns` works as expected:
```
# makedns sh-sl01
# host sh-sl01
sh-sl01.int has address 10.10.0.11
# host sh-sl01.ipmi
sh-sl01.ipmi is an alias for sh-sl01-bmc.ipmi.
sh-sl01-bmc.ipmi has address 10.11.0.11
```

In conclusion, we get an extra `<nodename>-<interface>` alias for the node, but in the end, a `<nodename>.<domain>` entry is generated in DNS, which is what we where looking for.

**Comment by bybai** [comment]
Hi @kcgthb ,
Your work around can resolve your problem. I guess there is no ip attribute in your node definition.
So there is only one line with sh-sl01 in /etc/hosts.   If there is ip attribute,  there will be 2 sh-sl01 in /etc/hosts in your example:
```
 ~]# lsdef testnode
 Object name: testnode
    groups=kvm,vm,all
    ip=10.10.0.51
    nicaliases.bmc=testnode
    nicips.bmc=20.11.0.11
    nicnetworks.bmc=net_ipmi
    postbootscripts=otherpkgs
    postscripts=syslog,remoteshell,syncfiles
 ~]# makehosts testnode
 ~]# cat /etc/hosts|grep testnode
 10.10.0.51 testnode testnode.cluster.com
 20.11.0.11 testnode-bmc testnode-bmc.ipmi testnode
```

In my example, there are 2 testnode in /etc/hosts, so the work around is not a general method for hostname in different domains.

**Comment by kcgthb** [comment]
Hi @bybai,

My bad, I left out the main interface to focus on the alternate `bmc` interface (with the `--nics` option of `lsdef`). But the node does have an ip attribute:
``` 
# lsdef sh-sl01 -i ip
Object name: sh-sl01
    ip=10.10.0.11
```
and the resulting `/etc/hosts` file contains two lines for that host, one for each interface, as you noted:
```
# grep sh-sl01 /etc/hosts
10.10.0.11 sh-sl01.int sh-sl01
10.11.0.11 sh-sl01-bmc.ipmi sh-sl01-bmc sh-sl01
```

There are 2 lines with the `sh-sl01` alias, but that's allows `makedns` to generate the correct entries on both domains. Having twice the same alias in `/etc/hosts` is fine, as long as the first one is in the default domain. And I believe that will always be the case, as `makehosts` generates the main interface entry first.

**Comment by bybai** [comment]
Hi @kcgthb ,
I think the following enhancements can  cover both issues #1086 and #2142:
After enhancements, makehosts will generate the default values in /etc/hosts as following:

If otherinterfaces is configured : -bmc:10.11.0.11, /etc/hosts is as :
```
10.10.0.11 sh-sl01.int sh-sl01 
10.11.0.11 sh-sl01-bmc.ipmi sh-sl01-bmc sh-sl01.ipmi
```

If otherinterfaces is configured: :10.11.0.11, /etc/hosts is as:
```
10.10.0.11 sh-sl01.int sh-sl01 
10.11.0.11 sh-sl01.ipmi
```

what is about your idea?

**Comment by kcgthb** [comment]
Looks interesting indeed, thanks!

**Comment by bybai** [comment]
Hi @kcgthb , we will start next release soon,  since my old pull request is closed, we plan to add this feature in next release plan. thanks.
